core/primitives/src/sharding.rs

```diff
@@ -98,20 +99,239 @@ pub struct ShardChunkHeader {
     pub hash: ChunkHash,
 }
 
+#[derive(BorshSerialize, BorshDeserialize, Serialize, Clone, PartialEq, Eq, Debug)]
+#[borsh_init(init)]
+pub struct ShardChunkHeaderV2 {
+    pub inner: ShardChunkHeaderInner,
+
+    pub height_included: BlockHeight,
+
+    /// Signature of the chunk producer.
+    pub signature: Signature,
+
+    #[borsh_skip]
+    pub hash: ChunkHash,
+}
+
+impl ShardChunkHeaderV2 {
+    pub fn init(&mut self) {
+        self.hash = Self::compute_hash(&self.inner);
+    }
+
+    pub fn compute_hash(inner: &ShardChunkHeaderInner) -> ChunkHash {
+        let inner_bytes = inner.try_to_vec().expect("Failed to serialize");
+        let inner_hash = hash(&inner_bytes);
+
+        let mut input_bytes: Vec<u8> = Vec::with_capacity(2 * inner_hash.as_ref().len());
```

Comment:
there is a utility function `combine_hash` that you can use here

Reply:
Done in https://github.com/nearprotocol/nearcore/pull/3463/commits/51733f96728cd4b73580376e3b6eb553d23e9442